### PR TITLE
Call captureStackTrace to filter boom from traces

### LINKED ISSUE #67: Use captureStackTrace to remove boom artifacts from stack traces
Using this https://code.google.com/p/v8-wiki/wiki/JavaScriptStackTraceApi the boom stack traces can safely remove their own frames so:

```
Debug: internal, implementation, error
    Error: child "text" fails because ["text" is required]
    at Object.exports.create (/Users/user/project/node_modules/boom/lib/index.js:21:17)
    at Object.exports.internal (/Users/user/project/node_modules/boom/lib/index.js:262:99)
    at Object.exports.badImplementation (/Users/user/project/node_modules/boom/lib/index.js:298:23)
    at postValidate (/Users/user/project/node_modules/hapi/lib/validation.js:194:26)
    at [object Object].internals.Any._validateWithOptions (/Users/user/project/node_modules/joi/lib/any.js:652:16)
    at [object Object].root.validate (/Users/user/project/node_modules/joi/lib/index.js:102:23)
    at exports.response (/Users/user/project/node_modules/hapi/lib/validation.js:213:20)
    at /Users/user/project/node_modules/hapi/lib/request.js:377:13
    at iterate (/Users/user/project/node_modules/hapi/node_modules/items/lib/index.js:35:13)
    at done (/Users/user/project/node_modules/hapi/node_modules/items/lib/index.js:27:25)
    at /Users/user/project/node_modules/hoek/lib/index.js:841:22
    at process._tickDomainCallback (node.js:381:11)
```

Could reduce to:

```
Debug: internal, implementation, error
    Error: child "text" fails because ["text" is required]
    at postValidate (/Users/user/project/node_modules/hapi/lib/validation.js:194:26)
    at [object Object].internals.Any._validateWithOptions (/Users/user/project/node_modules/joi/lib/any.js:652:16)
    at [object Object].root.validate (/Users/user/project/node_modules/joi/lib/index.js:102:23)
    at exports.response (/Users/user/project/node_modules/hapi/lib/validation.js:213:20)
    at /Users/user/project/node_modules/hapi/lib/request.js:377:13
    at iterate (/Users/user/project/node_modules/hapi/node_modules/items/lib/index.js:35:13)
    at done (/Users/user/project/node_modules/hapi/node_modules/items/lib/index.js:27:25)
    at /Users/user/project/node_modules/hoek/lib/index.js:841:22
    at process._tickDomainCallback (node.js:381:11)
```

Which helps remove noise from throws.

Potential downsides:
- Only works under v8
- Slight bit of additional complexity
- Won't impact `wrap` API (Probably not a downside since it's BYO `Error` object so the user controls the trace already

General pattern is here:
https://github.com/chaijs/chai/blob/4e18d2a49394f21f49eaea97f556d6a17ecbcc7e/chai.js#L5015
https://github.com/wycats/handlebars.js/blob/master/lib/handlebars/exception.js

Glad to do the work to PR this if desired.

### PR DESCRIPTION
**Pull request by kpdecker**
Fixes #67

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -259,7 +263,12 @@ exports.tooManyRequests = function (message, data) {
 
 exports.internal = function (message, data, statusCode) {
 
-    var error = (data instanceof Error ? exports.wrap(data, statusCode, message) : exports.create(statusCode || 500, message));
+    return internals.internal(message, data, statusCode, exports.internal);
+};
+
+internals.internal = function (message, data, statusCode, ctor) {
+
+    var error = (data instanceof Error ? exports.wrap(data, statusCode, message) : internals.create(statusCode || 500, message, ctor));
```
Comment (arb) [inline lib/index.js]: Can you rewrite this as an `if` `else`? I find executing functions inside ternary operators a little hard to follow for the humans.

File: lib/index.js [hunk 2]
```diff
@@ -259,7 +263,12 @@ exports.tooManyRequests = function (message, data) {
 
 exports.internal = function (message, data, statusCode) {
 
-    var error = (data instanceof Error ? exports.wrap(data, statusCode, message) : exports.create(statusCode || 500, message));
+    return internals.internal(message, data, statusCode, exports.internal);
+};
+
+internals.internal = function (message, data, statusCode, ctor) {
```
Comment (arb) [inline lib/index.js]: How about a more descriptive name? `generate` or `build` or something better that I can't think of.

**Comment by arb** [comment]
Do we want to expose something to make this behavior configurable?

File: lib/index.js (hunk 1 above)
Comment (kpdecker) [inline lib/index.js]: This is simply porting the existing code defined in the original method.

File: lib/index.js (hunk 2 above)
Comment (kpdecker) [inline lib/index.js]: This is the helper method to create internal errors. Could potentially call it `createInternal` or similar but on first look it seems like matching it to the external API of `exports.internal` makes more sense to keep parity with the API methods.

**Comment by kpdecker** [comment]
Regarding making this configurable, what is the use case for consumers of boom (vs. authors of boom)? From the consumer perspective I can't think of a case where they should care about the internals. From the author perspective, you do loose some potential data but this seems to be a reasonable price considering the relatively limited internal complexity of the library and the benefit provided to users.

**Comment by hueniverse** [comment]
Dropping the boom part of the stack makes sense to me. I don't think a setting it useful here.

File: lib/index.js (hunk 1 above)
Comment (arb) [inline lib/index.js]: Understood, but since you're in there anyway... :cake:

File: lib/index.js (hunk 2 above)
Comment (arb) [inline lib/index.js]: It's just not very descriptive of what it does. If someone wants to add another status code based boom type, they'll have to go looking to see what `internals.internal` does.

File: lib/index.js (hunk 1 above)
Comment (kpdecker) [inline lib/index.js]: Will update later tonight.

File: lib/index.js (hunk 2 above)
Comment (kpdecker) [inline lib/index.js]: `createInternal` then?

Comment (arb) [inline lib/index.js]: :shipit:

Comment (gergoerdosi) [inline lib/index.js]: I think `createInternal` is still confusing. 5xx errors are server errors. The word "internal" is only used for the 500 status code.

Comment (gergoerdosi) [inline lib/index.js]: Also, `internals.internal()` accepts arguments in different order than `internals.create()`. Is there a reason for that?

Since all calls to `internals.internal()` pass a status code, wouldn't be easier to just do everything in `internals.create()` by checking the status code?

**Comment by kpdecker** [comment]
Changed the `internal.internal` method to `internal.serverError`

**Comment by arb** [comment]
This looks great. How do you think this should be versioned? My gut tells me `major` since we are changing the shape of the error stack and someone could be relying on the way it used to work.

Thoughts @hueniverse?

**Comment by kpdecker** [comment]
The error stack's content was never api, right? Major seems unnecessary.
Minor or patch would be a toss up in my mind but I could argue for either
one and feel like it's an ok choice.

**Comment by AdriVanHoudt** [comment]
I would also go with minor or patch, not part of "public api" and would be weird to rely on the fact that it is in there imo

**Comment by hueniverse** [comment]
Minor.
